File: app/src/main/java/de/markusfisch/android/shadereditor/adapter/TextureAdapter.java
package de.markusfisch.android.shadereditor.adapter;

import android.content.Context;
import android.graphics.BitmapFactory;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import de.markusfisch.android.shadereditor.R;
import de.markusfisch.android.shadereditor.database.DataRecords.TextureInfo;

public class TextureAdapter extends BaseAdapter {
	private final String sizeFormat;
	private final List<TextureInfo> textures = new ArrayList<>();

	public TextureAdapter(@NonNull Context context) {
		sizeFormat = context.getString(R.string.texture_size_format);
	}

	@Override
	public int getCount() {
		return textures.size();
	}

	@Override
	public TextureInfo getItem(int position) {
		return textures.get(position);
	}

	@Override
	public long getItemId(int position) {
		return textures.get(position).id();
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		ViewHolder holder;
		if (convertView == null) {
			convertView = LayoutInflater.from(parent.getContext()).inflate(
					R.layout.row_texture, parent, false);
			holder = new ViewHolder(convertView);
			convertView.setTag(holder);
		} else {
			holder = (ViewHolder) convertView.getTag();
		}
		holder.bindTo(textures.get(position));
		return convertView;
	}

	public void setData(List<TextureInfo> newTextures) {
		this.textures.clear();
		if (newTextures != null) {
			this.textures.addAll(newTextures);
		}
		notifyDataSetChanged();
	}

	private class ViewHolder {
		private final ImageView preview;
		private final TextView name;
		private final TextView size;

		ViewHolder(View itemView) {
			preview = itemView.findViewById(R.id.texture_preview);
			name = itemView.findViewById(R.id.texture_name);
			size = itemView.findViewById(R.id.texture_size);
		}

		void bindTo(TextureInfo texture) {
			byte[] bytes = texture.thumb();
			if (bytes != null && bytes.length > 0) {
				preview.setImageBitmap(BitmapFactory.decodeByteArray(
						bytes, 0, bytes.length));
			} else {
				preview.setImageBitmap(null);
			}

			name.setText(texture.name());
			size.setText(String.format(sizeFormat, texture.width(), texture.height()));
		}
	}
}

File: app/src/main/java/de/markusfisch/android/shadereditor/adapter/TextureAdapter.kt
package de.markusfisch.android.shadereditor.adapter

import android.content.Context
import android.database.Cursor
import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CursorAdapter
import android.widget.ImageView
import android.widget.TextView
import de.markusfisch.android.shadereditor.R
import de.markusfisch.android.shadereditor.database.Database

open class TextureAdapter(
    context: Context, cursor: Cursor
) : CursorAdapter(context, cursor, false) {

    private val sizeFormat: String = context.getString(R.string.texture_size_format)

    private val nameIndex: Int = cursor.getColumnIndex(Database.TEXTURES_NAME)
    private val widthIndex: Int = cursor.getColumnIndex(Database.TEXTURES_WIDTH)
    private val heightIndex: Int = cursor.getColumnIndex(Database.TEXTURES_HEIGHT)
    private val thumbIndex: Int = cursor.getColumnIndex(Database.TEXTURES_THUMB)

    override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View {
        return LayoutInflater.from(context).inflate(R.layout.row_texture, parent, false)
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        val holder = getViewHolder(view)
        setData(holder, cursor)
    }

    fun getViewHolder(view: View): ViewHolder {
        return (view.tag as? ViewHolder) ?: ViewHolder(view).apply {
            view.tag = this
        }
    }

    fun setData(holder: ViewHolder, cursor: Cursor) {
        val bytes = cursor.getBlob(thumbIndex)
        val bitmap = bytes?.let {
            if (it.isNotEmpty()) {
                BitmapFactory.decodeByteArray(it, 0, it.size)
            } else null
        }
        holder.preview.setImageBitmap(bitmap)
        holder.name.text = cursor.getString(nameIndex)
        holder.size.text =
            String.format(sizeFormat, cursor.getInt(widthIndex), cursor.getInt(heightIndex))
    }

    class ViewHolder(view: View) {
        val preview: ImageView = view.findViewById(R.id.texture_preview)
        val name: TextView = view.findViewById(R.id.texture_name)
        val size: TextView = view.findViewById(R.id.texture_size)
    }
}
